agent/ebs/watcher_test.go

```diff
@@ -299,6 +299,125 @@ func TestHandleDuplicateEBSAttachment(t *testing.T) {
+			name: "StageAll returns expected error when CSI NodeStage call fails",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(2)
+			},
+			setCSIClientExpectations: func(mcc *mock_csiclient.MockCSIClient) {
+				mcc.EXPECT().NodeStageVolume(gomock.Any(),
+					taskresourcevolume.TestVolumeId, gomock.Any(),
+					filepath.Join(hostMountDir, taskresourcevolume.TestSourceVolumeHostPath),
+					taskresourcevolume.TestFileSystem, gomock.Any(), gomock.Any(),
+					gomock.Any(), gomock.Any(), gomock.Any()).Return(errors.New("test")).Times(1)
+			},
+			foundVolumes:             map[string]string{taskresourcevolume.TestVolumeId: taskresourcevolume.TestDeviceName},
+			expectedErrorStubs:       []string{"Failed to initialize EBS volume ID"},
+			expectedAttachmentStatus: attachment.AttachmentNone,
+		},
+		{
+			name: "StageAll returns nil when already attached",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ebsResourceAttachment.SetAttachedStatus()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(2)
+			},
+			foundVolumes: map[string]string{taskresourcevolume.TestVolumeId: taskresourcevolume.TestDeviceName},
+			// we won't call NodeStage here
+			setCSIClientExpectations: func(mcc *mock_csiclient.MockCSIClient) {},
+			expectedErrorStubs:       nil,
+			expectedAttachmentStatus: attachment.AttachmentAttached,
+		},
+		{
+			name: "StageAll fails with missing volume ID",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(nil, false).Times(1)
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(1)
+			},
+			foundVolumes:             map[string]string{taskresourcevolume.TestVolumeId: taskresourcevolume.TestDeviceName},
+			setCSIClientExpectations: func(mcc *mock_csiclient.MockCSIClient) {},
+			expectedErrorStubs:       []string{"Unable to find EBS volume with volume ID"},
+			expectedAttachmentStatus: attachment.AttachmentNone,
+		},
+		{
+			name: "StageAll multiple volumes returns a single error and continues if first volume stage fails",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ts.EXPECT().GetEBSByVolumeId("testVolId").Return(nil, false).Times(1)
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(2)
+			},
+			foundVolumes: map[string]string{"testVolId": "testDeviceName", taskresourcevolume.TestVolumeId: taskresourcevolume.TestDeviceName},
```

Comment:
because the map is unordered, we can't guarantee that the volumes are processed in order so this is the best I could come up with to test that we process beyond an early failure.  The test doesn't require ordering, just an output of a single attached volume and a single error where no volumes enter the case in attached state.